samcli/lib/utils/hash.py

```diff
@@ -0,0 +1,60 @@
+"""
+Hash calculation utilities for files and directories.
+"""
+import os
+import hashlib
+
+BLOCK_SIZE = 4096
+
+
+def file_checksum(file_name):
+    """
+
+    Parameters
+    ----------
+    file_name: file name of the file for which md5 checksum is required.
+
+    Returns
+    -------
+    md5 checksum of the given file.
+
+    """
+    with open(file_name, "rb") as file_handle:
+        md5 = hashlib.md5()
+
+        # Save current cursor position and reset cursor to start of file
+        curpos = file_handle.tell()
+        file_handle.seek(0)
+
+        buf = file_handle.read(BLOCK_SIZE)
+        while buf:
+            md5.update(buf)
+            buf = file_handle.read(BLOCK_SIZE)
+
+        # Restore file cursor's position
+        file_handle.seek(curpos)
+
+        return md5.hexdigest()
+
+
+def dir_checksum(directory, followlinks=True):
+    """
+
+    Parameters
+    ----------
+    directory : A directory with an absolute path
+    followlinks: Follow symbolic links through the given directory
+
+    Returns
+    -------
+    md5 checksum of the directory.
+
+    """
+    md5_dir = hashlib.md5()
```

Comment:
is md5 the right one? I am not sure if md5 is stable enough or has issues.

Reply:
for checksum purposes it should be fine (I dont think its cryptographically secure though), we can also look at sha256 as well.